src/wallet/sqlite.cpp

```diff
@@ -273,35 +278,40 @@ void SQLiteDatabase::Open()
+        if (!m_read_only) {
+            throw std::runtime_error("SQLiteDatabase: Database opened in readonly mode but read-write permissions are needed");
+        }
     }
 
-    // Acquire an exclusive lock on the database
-    // First change the locking mode to exclusive
-    SetPragma(m_db, "locking_mode", "exclusive", "Unable to change database locking mode to exclusive");
-    // Now begin a transaction to acquire the exclusive lock. This lock won't be released until we close because of the exclusive locking mode.
-    int ret = sqlite3_exec(m_db, "BEGIN EXCLUSIVE TRANSACTION", nullptr, nullptr, nullptr);
-    if (ret != SQLITE_OK) {
-        throw std::runtime_error("SQLiteDatabase: Unable to obtain an exclusive lock on the database, is it being used by another instance of " CLIENT_NAME "?\n");
-    }
-    ret = sqlite3_exec(m_db, "COMMIT", nullptr, nullptr, nullptr);
-    if (ret != SQLITE_OK) {
-        throw std::runtime_error(strprintf("SQLiteDatabase: Unable to end exclusive lock transaction: %s\n", sqlite3_errstr(ret)));
-    }
+    // Skip write operations when in read-only mode
+    if (!m_read_only) {
+        // Acquire an exclusive lock on the database
+        // First change the locking mode to exclusive
+        SetPragma(m_db, "locking_mode", "exclusive", "Unable to change database locking mode to exclusive");
+        // Now begin a transaction to acquire the exclusive lock. This lock won't be released until we close because of the exclusive locking mode.
+        int ret = sqlite3_exec(m_db, "BEGIN EXCLUSIVE TRANSACTION", nullptr, nullptr, nullptr);
+        if (ret != SQLITE_OK) {
+            throw std::runtime_error("SQLiteDatabase: Unable to obtain an exclusive lock on the database, is it being used by another instance of " CLIENT_NAME "?\n");
+        }
+        ret = sqlite3_exec(m_db, "COMMIT", nullptr, nullptr, nullptr);
+        if (ret != SQLITE_OK) {
+            throw std::runtime_error(strprintf("SQLiteDatabase: Unable to end exclusive lock transaction: %s\n", sqlite3_errstr(ret)));
+        }
 
-    // Enable fullfsync for the platforms that use it
-    SetPragma(m_db, "fullfsync", "true", "Failed to enable fullfsync");
+        // Enable fullfsync for the platforms that use it
+        SetPragma(m_db, "fullfsync", "true", "Failed to enable fullfsync");
 
-    if (m_use_unsafe_sync) {
-        // Use normal synchronous mode for the journal
-        LogPrintf("WARNING SQLite is configured to not wait for data to be flushed to disk. Data loss and corruption may occur.\n");
-        SetPragma(m_db, "synchronous", "OFF", "Failed to set synchronous mode to OFF");
+        if (m_use_unsafe_sync) {
+            // Use normal synchronous mode for the journal
+            LogPrintf("WARNING SQLite is configured to not wait for data to be flushed to disk. Data loss and corruption may occur.\n");
+            SetPragma(m_db, "synchronous", "OFF", "Failed to set synchronous mode to OFF");
+        }
```

Comment:
In f6f1f4199b4a3c2d7ef7c1bcfa122a9d8c5eeb64 "wallet: Allow read-only database access for info and dump commands"

These don't actually write anything to the database, they are runtime options. Removing the guard around these does not cause `info` or `dump` to fail.